pkg/apis/serving/v1beta1/route_validation.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/knative/pkg/apis"
+	"github.com/knative/serving/pkg/apis/serving"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+// Validate makes sure that Route is properly configured.
+func (r *Route) Validate(ctx context.Context) *apis.FieldError {
+	return serving.ValidateObjectMetadata(r.GetObjectMeta()).ViaField("metadata").Also(
+		r.Spec.Validate(withinSpec(ctx)).ViaField("spec")).Also(
+		r.Status.Validate(withinStatus(ctx)).ViaField("status"))
+}
+
+func validateTrafficList(ctx context.Context, traffic []TrafficTarget) *apis.FieldError {
+	var errs *apis.FieldError
+
+	// Track the targets of named TrafficTarget entries (to detect duplicates).
+	trafficMap := make(map[string]int)
+
+	sum := 0
+	for i, tt := range traffic {
```

Comment:
Since we're changing this, should we collect all? I.e. map[string][]int and then report error with all of them.

Reply:
Technically we do, just one pair at a time.  I expect the frequency that this happens with >1 will be infrequent enough that I'd rather not optimize for it.